src/nnpack/interface.jl

```diff
@@ -0,0 +1,100 @@
+    end
+end
+
+function conv(x::A1, w::A1, b::A2; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where {A1<:AbstractArray{Float32, 4}, A2<:AbstractArray{Float32, 1}}
+    pad_, stride_, fallback = check_support(x, (size(w, 1), size(w, 2)), pad, stride, dilation)
+    y = similar(x, cdims(size(x), dilation_dims(w, dilation), pad_, stride_))
+    if fallback
+        conv2d!(y, x, w, padding = pad, stride = stride, dilation = dilation)
+    else
+        conv!(y, x, w, b, pad = pad_, stride = stride_, dilation = dilation, algo = UInt32(algo))
+    end
+end
+
+function crosscor(x::A1, w::A1, b::A2; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where {A1<:AbstractArray{Float32, 4}, A2<:AbstractArray{Float32, 1}}
+    pad_, stride_, fallback = check_support(x, (size(w, 1), size(w, 2)), pad, stride, dilation)
+    y = similar(x, cdims(size(x), dilation_dims(w, dilation), pad_, stride_))
+    if fallback
+        conv2d!(y, x, w, padding = pad, stride = stride, dilation = dilation, mode = 1)
+    else
+        conv!(y, x, w, b, pad = pad_, stride = stride_, dilation = dilation, algo = UInt32(algo), flipkernel = 1)
+    end
+end
+
+function conv!(y::A1, x::A1, w::A1, b::A2; pad = 0, stride = 1, dilation = 1, algo = UInt32(0), flipkernel = 0) where {A1<:AbstractArray{Float32, 4}, A2<:AbstractArray{Float32, 1}}
+    if flipkernel == 0
+        w = reverse(reverse(w, dims=1), dims=2)
+    end
+    nnp_convolution_output(y, x, w, b, algo = algo, padding = pad, stride = stride)
+end
+
+crosscor!(y::A1, x::A1, w::A1, b::A2; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where {A1<:AbstractArray{Float32, 4}, A2<:AbstractArray{Float32, 1}} =
+    conv!(y, x, w, b, pad = pad, stride = stride, dilation = dilation, algo = algo, flipkernel = 1)
+
+∇conv_data(dy::A, x::A, w::A; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where A<:AbstractArray{Float64, 4} =
+    ∇conv_data(Float32.(dy), Float32.(x), Float32.(w), pad = pad, stride = stride, dilation = dilation, algo = algo)
+
+function ∇conv_data(dy::A, x::A, w::A; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where A<:AbstractArray{Float32, 4}
+    pad_, stride_, fallback = check_support(x, (size(w, 1), size(w, 2)), pad, stride, dilation)
+    if fallback
+        conv2d_grad_x!(zeros(Float32, size(x)), x, w, dy, padding = pad_, stride = stride_, dilation = dilation)
+    else
+        ∇conv_data!(zeros(Float32, size(x)), dy, x, w; pad = pad_, stride = stride_, dilation = dilation, algo = UInt32(algo))
+    end
+end
+
+function ∇conv_data!(dx::A, dy::A, x::A, w::A; pad = 0, stride = 1, dilation = 1, algo = UInt32(0), flipkernel = 0) where A<:AbstractArray{Float32, 4}
+    flipkernel == 0 && (w = reverse(reverse(w, dims=1), dims=2))
+    nnp_convolution_input_gradient(dx, x, dy, w, padding = pad, stride = stride, algo = algo)
+end
+
+function ∇conv_filter(dy::A, x::A, w::A; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where A<:AbstractArray{Float32, 4}
```

Comment:
I wonder about the fact that you're implementing both the in-place and out-of-place versions here. Couldn't you just add the in place ones, and NNlib's existing methods would dispatch the out-of-place ones?